File: web/js/clipPromptFixer.js
import { app } from "../../scripts/app.js";
import { builderInventory, mountPromptEditor } from "./promptEditor.js";

const NODE_NAME = "H3StudioClipPromptFixer";
const MIN_PLAN_HEIGHT = 120;
const SKILL_SLASH = "/prompt-minimax-h3-clip-fix";
const COPY_LABEL = "Copy skill command";
const COPY_TIP = "Copies /prompt-minimax-h3-clip-fix plus clip_index, plan, Builder dump, and the selected clips with one previous and one following clip.";

function targetNames(node) {
    return new Set([
        node?.type,
        node?.comfyClass,
        node?.constructor?.type,
        node?.constructor?.comfyClass,
        node?.constructor?.ComfyClass,
        node?.constructor?.nodeData?.name,
    ].filter(Boolean));
}

function isTarget(node) {
    return targetNames(node).has(NODE_NAME);
}

function findWidget(node, name) {
    return node?.widgets?.find((widget) => widget?.name === name);
}

function hidePlanWidget(widget) {
    if (!widget) return;
    widget.hidden = true;
    widget.serialize = true;
    widget.computeSize = () => [0, -4];
    widget.computedHeight = 0;
    if (widget.inputEl) widget.inputEl.style.display = "none";
    if (widget.element) widget.element.style.display = "none";
}

function linkedBuilder(node) {
    const slot = node.inputs?.find((item) => item?.name === "pack");
    if (!slot?.link || !node.graph) return null;
    const links = node.graph.links ?? node.graph._links;
    const link = typeof links?.get === "function" ? links.get(slot.link) : links?.[slot.link];
    const origin = node.graph.getNodeById?.(link?.origin_id ?? link?.[0]);
    if (!origin) return null;
    const names = new Set([
        origin.type, origin.comfyClass, origin.constructor?.type, origin.constructor?.comfyClass,
    ].filter(Boolean));
    return names.has("H3StudioBuilder") ? origin : null;
}

function isMusicVideoBuilder(builder) {
    return String(findWidget(builder, "mode")?.value || "") === "music_video";
}

function formatLinkedDump(builder) {
    if (!builder) return "";
    const items = builderInventory(builder);
    if (!items.length) return "";
    const music = isMusicVideoBuilder(builder);
    const lines = ["H3 Studio Builder pack"];
    const duration = Number(findWidget(builder, "max_clip_duration")?.value);
    if (Number.isFinite(duration) && duration > 0) lines.push(`duration: ${duration.toFixed(2)}s`);
    if (!music) {
        const segments = Number(findWidget(builder, "segments")?.value);
        if (Number.isFinite(segments) && segments > 0) lines.push(`segments: ${Math.round(segments)}`);
        if (findWidget(builder, "loop")?.value) lines.push("loop: true");
    }
    for (const item of items) {
        const desc = String(item.description || "").trim() || "(no description)";
        if (item.kind === "Model") {
            lines.push(`Model ${item.index}: ${desc}`);
        } else if (item.kind === "Picture") {
            const extra = item.media?.first_frame ? " (first frame)" : "";
            lines.push(`Picture ${item.index}: ${desc}${extra}`);
        } else if (item.kind === "Video") {
            const dur = Number(item.media?.duration) || 0;
            const extra = item.media?.has_soundtrack ? " (with soundtrack)" : "";
            lines.push(`Video ${item.index}: ${dur.toFixed(1)}s ${desc}${extra}`);
        } else if (item.kind === "Audio" && !music) {
            const dur = Number(item.media?.duration) || 0;
            lines.push(`Audio ${item.index}: ${dur.toFixed(1)}s ${desc}`);
        }
    }
    return `${lines.join("\n")}\n`;
}

function parseClipIndex(text) {
    const raw = String(text || "").trim().replace(/\s*-\s*/g, "-");
    if (!raw) return [];
    const found = [];
    const seen = new Set();
    for (const part of raw.split(/[,\s]+/)) {
        if (!part) continue;
        const match = part.match(/^(\d+)(?:-(\d+))?$/);
        if (!match) continue;
        let start = Number(match[1]);
        let end = match[2] != null ? Number(match[2]) : start;
        if (!(start >= 1) || !(end >= 1)) continue;
        if (end < start) [start, end] = [end, start];
        for (let n = start; n <= end; n++) {
            if (seen.has(n)) continue;
            seen.add(n);
            found.push(n);
        }
    }
    return found;
}

function windowIndices(selected) {
    if (!selected?.length) return [];
    const lo = Math.min(...selected) - 1;
    const hi = Math.max(...selected) + 1;
    const out = [];
    for (let n = Math.max(1, lo); n <= hi; n++) out.push(n);
    return out;
}

function extractPromptWindow(text, clipIndex) {
    const raw = String(text || "").trim();
    if (!raw) return "";
    const parsed = splitPromptClips(raw);
    const story = parsed.clips.filter((clip) => !clip.loop && clip.index >= 1);
    if (!story.length) return "";
    const available = new Set(story.map((clip) => clip.index));
    let selected = parseClipIndex(clipIndex);
    if (!selected.length) selected = story.map((clip) => clip.index);
    const want = new Set(windowIndices(selected).filter((n) => available.has(n)));
    const kept = story.filter((clip) => want.has(clip.index)).map((clip) => clip.text);
    return kept.length ? `${kept.join("\n\n").trim()}\n` : "";
}

function splitPromptClips(text) {
    const raw = String(text || "").replace(/\r\n/g, "\n").replace(/\r/g, "\n");
    const lines = raw.split("\n");
    const starts = [];
    for (let i = 0; i < lines.length; i++) {
        const stripped = lines[i].trim();
        const clip = stripped.match(/^(?:##\s+Clip\s+(\d+)\s*[—\-–]+|\*\*Clip\s+(\d+)\s*[—\-–]+)/i);
        if (clip) {
            starts.push({ line: i, index: Number(clip[1] || clip[2]), loop: false });
            continue;
        }
        if (/^(?:##\s+Loop\b|\*\*Loop\b)/i.test(stripped)) {
            starts.push({ line: i, index: null, loop: true });
        }
    }
    const header = starts.length ? lines.slice(0, starts[0].line).join("\n").trimEnd() : raw.trimEnd();
    const clips = starts.map((start, j) => {
        const end = j + 1 < starts.length ? starts[j + 1].line : lines.length;
        return {
            index: start.index,
            loop: start.loop,
            text: lines.slice(start.line, end).join("\n").trimEnd(),
        };
    });
    return { header, clips };
}

function formatSkillCommand(node) {
    const clipIndex = String(findWidget(node, "clip_index")?.value || "").trim();
    const plan = String(findWidget(node, "plan")?.value || "").trim();
    const seed = String(findWidget(node, "original_prompt")?.value || "").trim();
    const lines = [SKILL_SLASH];
    if (clipIndex) lines.push(`clip_index: ${clipIndex}`);
    lines.push("plan:");
    lines.push(plan || "(no extra plan)");
    lines.push("");
    const dump = formatLinkedDump(linkedBuilder(node));
    if (dump) {
        lines.push(dump.trimEnd());
        lines.push("");
    }
    lines.push("original_prompt:");
    lines.push(extractPromptWindow(seed, clipIndex).trimEnd());
    return `${lines.join("\n").replace(/\n+$/, "")}\n`;
}

async function copySkillCommand(node, button) {
    const text = formatSkillCommand(node);
    try {
        await navigator.clipboard.writeText(text);
    } catch (_) {
        window.prompt("Copy skill command", text);
    }
    button.textContent = "Copied";
    button.classList.add("h3-clip-fixer-copied");
    clearTimeout(button._h3Copied);
    button._h3Copied = setTimeout(() => {
        button.textContent = COPY_LABEL;
        button.classList.remove("h3-clip-fixer-copied");
    }, 1500);
}

function install(node) {
    if (!isTarget(node) || node.__h3ClipFixerUi) return;
    node.__h3ClipFixerUi = true;
    hidePlanWidget(findWidget(node, "plan"));
    const wrap = document.createElement("div");
    wrap.className = "h3-clip-fixer";
    wrap.innerHTML = `
<style>
.h3-clip-fixer {
  display: flex; flex-direction: column; gap: 6px; min-width: 0; min-height: 0; height: 100%;
  box-sizing: border-box;
}
.h3-clip-fixer-plan {
  position: relative; flex: 1 1 auto; min-height: ${MIN_PLAN_HEIGHT}px; min-width: 0;
  border: 1px solid #444; border-radius: 8px; overflow: hidden; background: #1c1c1c;
}
.h3-clip-fixer-plan .h3-studio-prompt-wrap {
  position: absolute; inset: 0; height: auto; width: auto;
}
.h3-clip-fixer-actions { display: flex; gap: 6px; flex-wrap: wrap; flex: 0 0 auto; align-items: center; }
.h3-clip-fixer-actions button {
  background: #333; color: #eee; border: 1px solid #555; padding: 4px 8px; cursor: pointer;
}
.h3-clip-fixer-actions button.h3-clip-fixer-copied { border-color: #6c9; color: #cfe; background: #2a4033; }
</style>
`;
    const host = document.createElement("div");
    host.className = "h3-clip-fixer-plan";
    const footer = document.createElement("div");
    footer.className = "h3-clip-fixer-actions";
    const copyBtn = document.createElement("button");
    copyBtn.type = "button";
    copyBtn.textContent = COPY_LABEL;
    copyBtn.title = COPY_TIP;
    copyBtn.addEventListener("pointerdown", (event) => {
        event.preventDefault();
        event.stopPropagation();
    });
    copyBtn.addEventListener("click", (event) => {
        event.preventDefault();
        event.stopPropagation();
        void copySkillCommand(node, copyBtn);
    });
    footer.appendChild(copyBtn);
    wrap.append(host, footer);
    const widget = node.addDOMWidget("h3_clip_fixer_plan", "div", wrap, {
        serialize: false,
        getMinHeight: () => MIN_PLAN_HEIGHT + 28,
        getHeight: () => "100%",
    });
    if (widget) widget.serialize = false;
    mountPromptEditor(node, host, {
        getValue: () => String(findWidget(node, "plan")?.value || ""),
        setValue: (text) => {
            const target = findWidget(node, "plan");
            if (target) target.value = text;
        },
        inventoryNode: node,
        label: "plan",
        placeholder: "What to change in the selected clips. Lyrics and timings stay locked.",
    });
    node.__h3GetInventory = () => {
        const builder = linkedBuilder(node);
        if (!builder) return [];
        return builderInventory(builder).map((item) => ({ ...item, builder }));
    };
    node.__h3FormatSkillCommand = () => formatSkillCommand(node);
}

app.registerExtension({
    name: "H3Studio.ClipPromptFixer",

    async nodeCreated(node) {
        if (!isTarget(node)) return;
        queueMicrotask(() => install(node));
    },

    loadedGraphNode(node) {
        if (!isTarget(node)) return;
        queueMicrotask(() => install(node));
    },
});

File: clip_prompt_fixer.py
"""Stash a seed prompt, clip indices, and a fix plan onto a Builder pack."""

from __future__ import annotations

import re

from .node_help import NODE_HELP
from .pack import require_pack
from .prompt_document import parse_prompt_document, story_clips

PROMPT_MODE_CLIP_FIX = "clip_fix"
_INDEX_TOKEN = re.compile(r"^(\d+)(?:-(\d+))?$")


def parse_clip_index(text: str) -> list[int]:
    raw = re.sub(r"\s*-\s*", "-", str(text or "").strip())
    if not raw:
        return []
    found = []
    seen = set()
    for part in re.split(r"[,\s]+", raw):
        if not part:
            continue
        match = _INDEX_TOKEN.match(part)
        if not match:
            raise ValueError(f"h3_studio: invalid clip_index {part!r}")
        start = int(match.group(1))
        end = int(match.group(2)) if match.group(2) is not None else start
        if start < 1 or end < 1:
            raise ValueError("h3_studio: clip_index must be >= 1")
        if end < start:
            start, end = end, start
        for n in range(start, end + 1):
            if n not in seen:
                seen.add(n)
                found.append(n)
    return found


def seed_is_partial(parsed) -> bool:
    story = story_clips(parsed)
    indices = [int(clip["index"]) for clip in story]
    if not indices:
        raise ValueError("h3_studio: prompt has no ## Clip sections")
    n = len(indices)
    header_seg = parsed.get("segments")
    if header_seg is not None and int(header_seg) != n:
        return True
    return sorted(indices) != list(range(1, n + 1))


def resolve_fix_clips(clip_index, parsed) -> list[int]:
    story = story_clips(parsed)
    available = {int(clip["index"]) for clip in story}
    if not available:
        raise ValueError("h3_studio: prompt has no ## Clip sections")
    requested = parse_clip_index(clip_index)
    if not requested:
        if not seed_is_partial(parsed):
            raise ValueError(
                "h3_studio: clip_index is required when original_prompt has a full clip set"
            )
        return [int(clip["index"]) for clip in story]
    missing = [n for n in requested if n not in available]
    if missing:
        raise ValueError(f"h3_studio: original_prompt has no clip {missing[0]}")
    return requested


_CLIP_HEAD = re.compile(
    r"^(?:##\s+Clip\s+(\d+)\s*[—\-–]+|\*\*Clip\s+(\d+)\s*[—\-–]+)",
    re.I,
)
_LOOP_HEAD = re.compile(r"^(?:##\s+Loop\b|\*\*Loop\b)", re.I)


def window_indices(selected) -> list[int]:
    nums = [int(n) for n in (selected or [])]
    if not nums:
        return []
    lo = min(nums) - 1
    hi = max(nums) + 1
    return list(range(max(1, lo), hi + 1))


def extract_prompt_window(text: str, clip_index: str) -> str:
    """Selected story clips plus one previous and one following. No document header."""
    raw = str(text or "").replace("\r\n", "\n").replace("\r", "\n")
    if not raw.strip():
        return ""
    lines = raw.split("\n")
    starts = []
    for i, line in enumerate(lines):
        stripped = line.strip()
        match = _CLIP_HEAD.match(stripped)
        if match:
            starts.append((i, int(match.group(1) or match.group(2)), False))
            continue
        if _LOOP_HEAD.match(stripped):
            starts.append((i, None, True))
    if not starts:
        return ""
    clips = []
    for j, (start, index, is_loop) in enumerate(starts):
        end = starts[j + 1][0] if j + 1 < len(starts) else len(lines)
        block = "\n".join(lines[start:end]).rstrip()
        clips.append((index, is_loop, block))
    story = [(index, block) for index, is_loop, block in clips if not is_loop and index]
    available = {index for index, _block in story}
    selected = parse_clip_index(clip_index)
    if not selected:
        selected = [index for index, _block in story]
    want = set(window_indices(selected)) & available
    kept = [block for index, block in story if index in want]
    if not kept:
        return ""
    return "\n\n".join(kept).strip() + "\n"


class H3StudioClipPromptFixer:
    @classmethod
    def INPUT_TYPES(cls):
        return {
            "required": {
                "pack": ("H3_STUDIO_PACK", {
                    "tooltip": (
                        "Wire H3 Studio Builder here. Inventory, song, lyrics, and duration "
                        "stay on the pack; this node only adds clip-fix metadata."
                    ),
                }),
                "original_prompt": ("STRING", {
                    "multiline": True, "default": "", "dynamicPrompts": False,
                    "tooltip": (
                        "Paste the Music Video or Auto Chain prompt. Full document or just "
                        "the ## Clip sections to rewrite."
                    ),
                }),
                "clip_index": ("STRING", {
                    "default": "",
                    "tooltip": (
                        "Story clip indices to rewrite, e.g. 11-12 or 11,12. Empty is allowed "
                        "only when original_prompt is a partial paste."
                    ),
                }),
                "plan": ("STRING", {
                    "multiline": True, "default": "", "dynamicPrompts": False,
                    "tooltip": (
                        "What to change in the selected clip bodies. Lyrics and timings stay "
                        "locked. Overrides the Builder plan for Local Prompter."
                    ),
                }),
            },
        }

    RETURN_TYPES = ("H3_STUDIO_PACK",)
    RETURN_NAMES = ("pack",)
    OUTPUT_TOOLTIPS = (
        "Same Builder pack plus clip_fix metadata for Local Prompter.",
    )
    FUNCTION = "attach_fix"
    CATEGORY = "H3 Studio"
    DESCRIPTION = NODE_HELP["H3StudioClipPromptFixer"]

    def attach_fix(self, pack, original_prompt, clip_index="", plan=""):
        pack = dict(require_pack(pack))
        plan_text = str(plan or "").strip()
        if not plan_text:
            raise ValueError("h3_studio: Plan is required")
        seed = str(original_prompt or "").strip()
        if not seed:
            raise ValueError("h3_studio: original_prompt is empty")
        parsed = parse_prompt_document(seed)
        pack["plan"] = plan_text
        pack["prompt_mode"] = PROMPT_MODE_CLIP_FIX
        pack["seed_prompt"] = seed
        pack["fix_clips"] = resolve_fix_clips(clip_index, parsed)
        return (pack,)
